src/commands/DeployVDBCommand.ts

```diff
@@ -14,17 +14,96 @@
+import * as kubectlapi from 'vscode-kubernetes-tools-api';
+import { DVTreeItem } from '../model/tree/DVTreeItem';
+import { EnvironmentVariableRefTreeNode } from '../model/tree/EnvironmentVariableRefTreeNode';
+import { DataSourceRefTreeNode } from '../model/tree/DataSourceRefTreeNode';
 
-export function deployVDBCommand(ctx) {
-	if (ctx instanceof DVProjectTreeNode) {
-		const prjNode: DVProjectTreeNode = ctx;
-		const file: string = prjNode.file;
-		handleDeploy(file);
+export async function deployCommand(prjNode: DVProjectTreeNode) {
+	// determine configmaps / secrets to deploy first
+	await handleReferences(prjNode, true);
+	await handleDeploy(prjNode.dvConfig.metadata.name, prjNode.file);
+}
+
+async function handleDeploy(name: string, file: string) {
+	const k8sApi: kubectlapi.API<kubectlapi.KubectlV1> = await kubectlapi.extension.kubectl.v1;
+	if (k8sApi && k8sApi.available) {
+		try {
+			let res: kubectlapi.KubectlV1.ShellResult;
+			const alreadyDeployed: boolean = await utils.isVDBDeployed(name);
+			if (alreadyDeployed) {
+				res = await k8sApi.api.invokeCommand(`replace -f ${file}`);
+			} else {
+				res = await k8sApi.api.invokeCommand(`create -f ${file}`);
+			}
+			if (res.code === 0) {
+				extension.log(`${res.stdout}`);
+				vscode.window.showInformationMessage(`Deployment of ${file} succeeded. Check the output view for more details.`);
+			} else {
+				extension.log(`${res.stderr}`);
+				vscode.window.showErrorMessage(`Deployment of ${file} failed. Please check the output view for more details.`);
+			}
+		} catch (error) {
+			extension.log(error);
+			vscode.window.showErrorMessage(`Deployment of ${file} failed. Please check the output view for more details.`);
+		}
+	} else {
+		extension.log(`Unable to acquire Kubernetes API. Make sure you have configured Kubernetes correctly and you are logged in.`);
+		vscode.window.showErrorMessage(`Unable to acquire Kubernetes API. Make sure you have configured Kubernetes correctly and you are logged in.`);
 	}
 }
 
-function handleDeploy(filepath: string): void {
-	extension.log(`DEPLOY: Selected File: ${filepath}`);
+export async function handleReferences(prjNode: DVTreeItem, deploy: boolean) {
+	let name: string;
+	let file: string;
+	let type: string;
+
+	for (let element of prjNode.getProject().getDataSourcesNode().children) {
+		if (element instanceof DataSourceRefTreeNode) {
+			const refDS: DataSourceRefTreeNode = element;
+			name = refDS.getReferenceName();
+			file = utils.getFullReferenceFilePath(prjNode.getProject().file, refDS.getReferenceName());
+			type = refDS.getReferenceType();
+			if (name && file && type) {
+				try {
+					const exists : boolean = await utils.isResourceDeployed(name, type);
+					if (!exists && deploy) {
+						await utils.deployResource(file, type);
+					} else if (exists && deploy) {
+						await utils.redeployResource(file, type);
+					} else if (exists && !deploy) {
+						await utils.undeployResource(name, type);
```

Comment:
the case of `!exists && !deploy` is silently doing nothing. it would be better to notify user that the resource cannot be undeployed (or that the menu is not available if there is nothing to undeploy)

Reply:
it will be done in a separate task. already discussed that with Ramesh yesterday and we will probably end up changing the element type of a deployed vdb so we can filter the undeploy for not deployed vdb files. Haven't had the time to create the jira though.

Reply:
>     * I have an error when trying to use a configmap for which I give a name conatining MAJ: `Unable to create ConfigMap /home/apupier/ws/camelk-test/MYCONFIGMAP.yaml.`
>       content of the configmap:
> 
> 
> ```
> apiVersion: v1
> kind: ConfigMap
> metadata:
>   name: MYCONFIGMAP
> data:
>   AVAR: witrhvalue
> ```
> 
> The file /home/apupier/ws/camelk-test/MYCONFIGMAP.yaml is created but it is not deployed

I have to admit I don't understand what you did and where you see the issue. Just creating a configmap doesn't deploy it. You have to use that configmap from inside your vdb and then deploy the vdb to finally have the configmap deployed.

Reply:
your last comment is at a wrong place.



> it will be done in a separate task. already discussed that with Ramesh yesterday and we will probably end up changing the element type of a deployed vdb so we can filter the undeploy for not deployed vdb files and also add some visual hint so users know if its deployed yet or not. Haven't had the time to create the jira though.

would be worthy that for this iteration there is at least a log mentioning that there is no undeploy because the resource is nto deployed.

Reply:
yep, agreed